lmod_manager/lmod_manager.py

```diff
@@ -126,5 +126,125 @@ def install(args: argparse.Namespace) -> Union[int, str]:
+        return f"directory '{installation_dir}' seems not to contain a valid installation"
+
+    config_file = _config_file(args.lmod_modules_dir, name, version)
+
+    if not config_file.exists():
+        return f"config file '{config_file}' not found"
+
+    shutil.rmtree(installation_dir)
+    config_file.unlink()
+
+    return 0
+
+
 def assert_never(value: NoReturn) -> NoReturn:
     assert False, f"Unhandled value: {value} ({type(value).__name__})"
+
+
+def _archive_type(name: str) -> Type:
+    if name.startswith("gnatpro"):
+        return Type.GNAT
+    if name.startswith("spark-pro"):
+        return Type.SPARK
+    if name.startswith("codepeer"):
+        return Type.CODEPEER
+    raise ValueError
+
+
+def _archive_name(module_type: Type) -> str:
+    if module_type is Type.GNAT:
+        return "gnatpro"
+    if module_type is Type.SPARK:
+        return "spark-pro"
+    if module_type is Type.CODEPEER:
+        return "codepeer"
+    assert_never(module_type)  # pragma: no cover
+
+
+def _module_type(name: str) -> Type:
+    if name == "gnatpro":
+        return Type.GNAT
+    if name == "sparkpro":
+        return Type.SPARK
+    if name == "codepeer":
+        return Type.CODEPEER
+    raise ValueError
+
+
+def _module_name(module_type: Type) -> str:
+    if module_type is Type.GNAT:
+        return "gnatpro"
+    if module_type is Type.SPARK:
+        return "sparkpro"
+    if module_type is Type.CODEPEER:
+        return "codepeer"
+    assert_never(module_type)  # pragma: no cover
+
+
+def _archive_regex(archive_type: Type) -> str:
+    archive_name = _archive_name(archive_type)
+    return rf"{archive_name}-([\d.w]*(?:-\d*)?)-([\w-]*)-(linux(?:64|)?)-bin\.tar\.gz"
+
+
+def _extracted_archive_dir(archive_type: Type, version: str, target: str, linux: str) -> Path:
+    archive_name = _archive_name(archive_type)
+    return Path(f"{archive_name}-{version}-{target}-{linux}-bin")
+
+
+def _install_archive(
+    archive_type: Type, installation_dir: Path, extracted_archive_dir: Path
+) -> None:
+    if archive_type is Type.GNAT:
+        call(f"cd {extracted_archive_dir} && ./doinstall {installation_dir}", shell=True)
+    elif archive_type is Type.SPARK:
+        call(f"echo '{installation_dir}' | {extracted_archive_dir}/doinstall", shell=True)
+    elif archive_type is Type.CODEPEER:
+        call(f"cd {extracted_archive_dir} && ./doinstall {installation_dir}", shell=True)
+    else:
+        assert_never(archive_type)  # pragma: no cover
+
+
+def _installation_dir(prefix: Path, name: str, version: str) -> Path:
+    return prefix / name / version
+
+
+def _installation_file(module_type: Type) -> Path:
+    if module_type is Type.GNAT:
+        return Path("bin/gnat")
+    if module_type is Type.SPARK:
+        return Path("bin/gnatprove")
+    if module_type is Type.CODEPEER:
+        return Path("bin/codepeer")
+    assert_never(module_type)  # pragma: no cover
```

Comment:
I've seen multiple of these if clauses, maybe it makes sense to create an abstract type class with properties that are the implemented for GNAT, SPARK and Codepeer?

Reply:
You are right, I have refactored the code.